### ISSUE TITLE
Sorting with proofs

### ISSUE THREAD

**Issue by siddhartha-gadgil** [siddhartha-gadgil/LTS2019#2]
Implement sorting algorithms such as _quick sort_ or _merge sort_ but with _proofs_, i.e.
* Define a `Sorted(v)` type parametrized by a vector `v`(the _input_) that corresponds to
   - an output vector
   - a proof that the output is sorted
   - a proof that the input and output differ by a permutation.
* Define sorting algorithms as dependent functions mapping `v` to `Sorted(v)`

**Comment by siddhartha-gadgil** [comment]
Yes, ideally not in a new folder but just in `Code`

**Comment by anotherArka** [comment]
Do we sort vectors of natural number or a general ordered type?

**Comment by siddhartha-gadgil** [comment]
A general ordered type would be best, but should be easy to upgrade from
natural numbers.

**Comment by AR-MA210** [comment]
Are other sorting algorithms allowed (other than Quick Sort and Merge Sort)?

Also, as the project is collaborative, is it okay to submit code in pieces?

- Abishek

**Comment by siddhartha-gadgil** [comment]
Yes, all algorithms are welcome. But a good sorting one like quick sort gets more credit than say _insertion sort_ - though it may be good to start with something simple.

It is good to submit in pieces, but it should be clear how the piece fits in the planned whole.

**Comment by anotherArka** [comment]
The problem with a general ordered type is we don't always have a proof of the ordering similar to LTE for Nat.

**Comment by shafilmaheenn** [comment]
Sir, the code that i gave for the merge function was actually wrong. I am really sorry.
Can you please remove it

**Comment by siddhartha-gadgil** [comment]
Just delete it and send a fresh pull request, making sure you pull from my
branch before editing.

Regards
Siddhartha

**Comment by shafilmaheenn** [comment]
I have corrected the code and sent a new pull request

**Comment by shafilmaheenn** [comment]
I think that the merge function that I gave doesnt work in linear time as it is supposed to (because of the sillycast function that I had to introduce).Can someone else please confirm this
